nemo_curator/modules/semdedup/clustering.py

```diff
@@ -0,0 +1,147 @@
+
+    ddf = dask_cudf.read_parquet(
+        emb_pqt_loc,
+        columns=["embeddings", args.id_col["name"]],
+        split_row_groups=False,
+    )
+    logger.info(f"Total number of partitions in input_ddf: {ddf.npartitions}")
+    # Persist ddf to save IO costs in host memory, should be able to do this via
+    # spilling to (TODO)
+    ddf = ddf.to_backend("pandas").persist()
+    ddf = ddf.repartition(npartitions=num_workers * 4)
+    wait(ddf)
+    client.rebalance(ddf)
+    # Switch back to GPU
+    ddf = ddf.to_backend("cudf")
+
+    cupy_darr = ddf.map_partitions(get_embedding_ar, meta=cp.ndarray([1, emb_size]))
+
+    cupy_darr.compute_chunk_sizes()
+
+    kmeans = KMeans(n_clusters=ncentroids, max_iter=niter, oversampling_factor=10)
+    logger.info("KMeans starting fit")
+    kmeans.fit(cupy_darr)
+    logger.info("KMeans fit complete")
+
+    logger.info("Computing nearest centroids using kmeans.predict")
+    nearest_cents = kmeans.predict(cupy_darr)
+    ddf["nearest_cent"] = nearest_cents.astype(np.int32)
+    logger.info("Nearest centroids computed")
+
+    logger.info("Computing distances to centroids using add_dist_to_cents")
+    meta_df = ddf._meta.copy()
+    meta_df["dist_to_cent"] = cp.zeros(1)
+    ddf = ddf.map_partitions(
+        add_dist_to_cents, centriods=kmeans.cluster_centers_, meta=meta_df
+    )
+    logger.info("Distances to centroids computed")
+
+    centroids = kmeans.cluster_centers_
+    logger.info("Centroids computed")
+
+    ddf = ddf.reset_index(drop=True)
+    return centroids, ddf
+
+
+if __name__ == "__main__":
+    # Configure command line arguments
+    args = parse_arguments()
+
+    # Kmeans can only be done with L2 using cuML.
+    assert args.clustering["Kmeans_with_cos_dist"] == False
+
+    # TODO: Cleanup below
+    save_folder = f'{args.root}/{args.clustering["save_loc"]}'
+    os.makedirs(save_folder, exist_ok=True)
+
+    # Initialize logger
+    log_file = os.path.join(save_folder, "compute_centroids.log")
+    logger = get_logger(
+        file_name=log_file,
+        level=logging.INFO,
+        stdout=True,
+    )
+    with open(pathlib.Path(save_folder, "clustering_params.txt"), "w") as fout:
+        pprint.pprint(args, fout)
+
+    kmeans_file_loc = pathlib.Path(save_folder, "kmeans_centroids.npy")
+
+    # Initialize dask client
+    client = get_client(**parse_client_args(args))
+    dt1 = datetime.now()
+    print("Start time:", dt1)
+    centroids, ddf = compute_centroids(args, logger, client)
+    ddf.to_parquet(
+        f"{save_folder}/embs_by_nearest_center/",
+        index=False,
+        partition_on="nearest_cent",
+    )
+    del ddf
+    kmeans_centroids_file = pathlib.Path(save_folder, "kmeans_centroids.npy")
+    np.save(kmeans_centroids_file, centroids)
+    dt2 = datetime.now()
+    elapse = dt2 - dt1
+    print("End time:", dt2)
+    print("elapse:", elapse)
+
+    client.cancel(client.futures, force=True)
+    client.close()
```

Comment:
I'm pretty sure it doesn't have an impact, but would be good to confirm that this doesn't impact externally created dask clusters.

Reply:
I added it because `cuML` cluster shutdown is noisy and this helps decrease the noise.